### PR TITLE
First PR of model using toy data

### PR DESCRIPTION
**Pull request by nrosed**
In this PR there are many files changed, but most of them are different experiments from developing this method. 

The file that requires review is: `./experiments/rotated_example_ssDIVA_mnist.ipynb` 

This notebook is self-contained, so it creates the data, builds the model, trains, and validates. I am very rusty with python and still getting used to Keras, so please feel free to identify any coding practices I am doing incorrectly or tips on improving the code. It would be greatly appreciated :). The code is broken down into the following sections:

- Intro
- Github / VM setup
- Imports
- Load / Pre-process data
- Model Definition
    - Hyperparameters
    - Encoder
    - Latent Space
    - Decoder + Classifiers
    - Loss
- Make the computational graph
- Experimental Setup
- Training
- Inspect the output
    - TSNE unseen data (75 degrees)
    - TSNE unlabeled data (30 degrees)
    - TSNE All data (0-75)
    - TSNE Training data
- Generate Data
    - Walk across latent space
- Predict

The **intro** section has a description of how the model works and the intuition behind it, but let me know if anything is not clear.

If you would like to interact with the code or have a better interface with the code I think you can run it here: https://colab.research.google.com/drive/1qWnDztzY2nScl9RoBHwF2tkd7YcDwGiB?usp=sharing

### CODE REVIEW

**Review by ben-heil — APPROVED** [review]
0) Domain invariance/OOD learning is really cool and imo the most important problem in DL. I'm really excited that someone in the lab is working on it!
1) Since your IDE saves notebooks on a single line, can you add the converted python version of the notebooks to make attaching comments to code easier? You can generate the converted versions with `jupyter nbconvert --to python <notebook_path>`. Alternatively CoLab has a download as .py button if you're using CoLab for your IDE
2) Cell 56, what is the purpose of the RELU layer getting called on z_rot? Doesn't that just chop off the negative half of the Gaussian? I probably don't fully understand the classifier architecture though, or I'm missing something about Keras
3) Just a heads up, I haven't used much Keras since undergrad so I'm not great at reading the syntax. Most people in the lab use pytorch since it's easier to debug. Not that you have to switch since you already have a lot of code written, but if you're already writing your own training loops like in cell 61 then Keras isn't saving you much work
4) As your models take longer to train, you may want to use something like tensorboard or [weights and biases](https://docs.wandb.ai/guides/integrations) to visualize training progress
5) The figure in cell 77 is so cool!
6) Is there a reason that the 5 and 7 classes have such low classification accuracy in cell 80?

**Comment by nrosed** [comment]
> Domain invariance/OOD learning is really cool and imo the most important problem in DL. I'm really excited that someone in the lab is working on it!

yay! Hopefully it proves fruitful...

> Since your IDE saves notebooks on a single line, can you add the converted python version of the notebooks to make attaching comments to code easier? You can generate the converted versions with jupyter nbconvert --to python <notebook_path>. Alternatively CoLab has a download as .py button if you're using CoLab for your IDE

ah thanks for letting me know! I added the `.py` file in https://github.com/greenelab/sc_bulk_ood/pull/1/commits/5250406c0dc9eae3c45870d2c7fac96e88978f2b.


> Cell 56, what is the purpose of the RELU layer getting called on z_rot? Doesn't that just chop off the negative half of the Gaussian? I probably don't fully understand the classifier architecture though, or I'm missing something about Keras

hmm, yeah. I don't really have a justification. I will play around with different activation layers. Added a note in https://github.com/greenelab/sc_bulk_ood/issues/2


> Just a heads up, I haven't used much Keras since undergrad so I'm not great at reading the syntax. Most people in the lab use pytorch since it's easier to debug. Not that you have to switch since you already have a lot of code written, but if you're already writing your own training loops like in cell 61 then Keras isn't saving you much work

Ah, ok good to know. I was actually thinking about using pytorch+pyro, because implementing priors is really confusing to me in Keras since it has to interface with TensorFlow probability layers.

> As your models take longer to train, you may want to use something like tensorboard or weights and biases to visualize training progress

Yes -- I plan to. I had it working in another notebook, but removed it in this one because it wasn't working correctly with the custom training loop. Noted in 

> The figure in cell 77 is so cool!

yay!

> Is there a reason that the 5 and 7 classes have such low classification accuracy in cell 80?

No, this varies a lot with each training run. I'm not doing a proper hyperparameter search currently, so hopefully the performance will improve and also become more stable.
